File: web/src/domain/entities/document.ts
// 与后端 domain/document 对齐;前端只用到展示层需要的字段。
export type DocumentSource = 'note' | 'knowledge';
export type DocumentContentType = 'markdown' | string;
// 0=dirty, 1=chunked, 2=error
export type DocumentChunkStatus = 0 | 1 | 2;

export interface Document {
  id: number;
  title: string;
  source: DocumentSource;
  lang: string;
  content_type: DocumentContentType;
  chunk_status: DocumentChunkStatus;
  current_version_id: number | null;
  archived_at: string | null;
  created_at: string;
  updated_at: string | null;
}

// 详情 — 列表返回 Document,详情额外带 raw content。
export interface DocumentDetail extends Document {
  content: string;
}

// content 必填,lang / content_type / parent_tree_id 可空;parent_tree_id 仅 knowledge 生效。
export interface CreateDocumentInput {
  source: DocumentSource;
  title: string;
  lang?: string;
  content_type?: DocumentContentType;
  parent_tree_id?: number | null;
  content: string;
}

export interface CreateDocumentResult {
  document_id: number;
}

export interface ListDocumentsInput {
  source?: DocumentSource;
  chunk_status?: DocumentChunkStatus;
  include_archived?: boolean;
  limit?: number;
  offset?: number;
}

export interface ListDocumentsResult {
  items: Document[];
  total: number;
}

// 文档移动入参;parent_tree_id 可空(根级)。
export interface MoveDocumentInput {
  parent_tree_id: number | null;
}

File: internal/domain/document/document.go
package document

import (
	"time"

	"learn/internal/domain"
)

type ChunkStatus int8

const (
	ChunkStatusDirty   ChunkStatus = 0 // pending
	ChunkStatusChunked ChunkStatus = 1 // chunked; retrievable
	ChunkStatusError   ChunkStatus = 2 // split failed
)

type ContentType string

const (
	ContentTypeMarkdown ContentType = "markdown"
	ContentTypePDF      ContentType = "pdf"
)

const defaultContentType = ContentTypeMarkdown

func ValidContentTypes() map[ContentType]bool {
	return map[ContentType]bool{
		ContentTypeMarkdown: true,
		ContentTypePDF:      true,
	}
}

func NormalizeContentType(ct ContentType) ContentType {
	if ct == "" {
		return defaultContentType
	}
	return ct
}

func ValidateContentType(ct ContentType) error {
	if ct == "" {
		return nil
	}
	if !ValidContentTypes()[ct] {
		return domain.ErrDocumentInvalidContentType
	}
	return nil
}

// Notes allow markdown only; knowledge any whitelisted.
func ValidateSourceContentMatch(source string, ct ContentType) error {
	if err := ValidateContentType(ct); err != nil {
		return err
	}
	if source == "note" && ct != ContentTypeMarkdown {
		return domain.ErrDocumentNoteMarkdownOnly
	}
	return nil
}

type CreateParams struct {
	Source       string // "note" | "knowledge"
	Title        string
	Lang         string      // optional, default "zh"
	ContentType  ContentType // optional, default markdown
	ParentTreeID *int64      // knowledge only
	Content      string      // required
}

// Location lives in knowledge_tree adjacency table.
type Document struct {
	ID               int64
	CurrentVersionID *int64
	UserID           int64
	Source           string
	Title            string
	Lang             string
	ContentType      ContentType
	ChunkStatus      ChunkStatus
	ArchivedAt       *time.Time
	CreatedAt        time.Time
	UpdatedAt        *time.Time
}
